### PR TITLE
feat: Update the Pydantic from v1 to v2

### LINKED ISSUE #3778: Pydantic 2.0 support
**Is your feature request related to a problem? Please describe.**
I need to install `feast` with Pydantic 2.0 support, but it's pinned to `pydantic>=1,<2`

**Describe the solution you'd like**
Update dependency to `pydantic>=2`

Pydantic v2 offers improved performance (2x-50x faster), which can contribute to the overall efficiency of Feast.

Updating Pydantic to version 2 will be a valuable enhancement to the `feast`, and I am willing to contribute to this update if needed. I look forward to hearing your thoughts on this proposal and am open to discussing it further.

### PR DESCRIPTION
**Pull request by shuchu**
**What this PR does / why we need it**:
Based on the requirements from the Feast users, we want to upgrade the Pydantic version to v2 (2.6) for the performance of Feast.

Here are a few important package version changes:
pydantic >= 2.0.0
great_expectations>=0.15.41 (the upper bound is removed. Its required by the pydantic v2.)

More details can refer to the previous PR:  https://github.com/feast-dev/feast/pull/3942

**Which issue(s) this PR fixes**:
Fixes #3778

### CODE REVIEW

File: sdk/python/feast/infra/offline_stores/contrib/athena_offline_store/tests/data_source.py [hunk 1]
```diff
@@ -43,15 +43,16 @@ def __init__(self, project_name: str, *args, **kwargs):
             workgroup=workgroup,
             s3_staging_location=f"s3://{bucket_name}/test_dir",
         )
+        self,
```
Comment (sudohainguyen) [inline sdk/python/feast/infra/offline_stores/contrib/athena_offline_store/tests/data_source.py:46]: is this typo?

File: sdk/python/feast/infra/passthrough_provider.py [hunk 2]
```diff
@@ -70,7 +70,7 @@ def batch_engine(self) -> BatchMaterializationEngine:
         if self._batch_engine:
             return self._batch_engine
         else:
-            engine_config = self.repo_config._batch_engine_config
+            engine_config = self.repo_config.batch_engine_config
```
Comment (sudohainguyen) [inline sdk/python/feast/infra/passthrough_provider.py:73]: double check this change, should be `_batch_engine_config`

File: sdk/python/feast/infra/offline_stores/snowflake_source.py [hunk 3]
```diff
@@ -279,20 +280,21 @@ def get_table_column_names_and_types(
...
                     row["snowflake_type"] = "NUMBERwSCALE"
 
-            elif row["type_code"] in [5, 9, 12]:
+            elif row["type_code"] in {5, 9, 12}:
                 error = snowflake_unsupported_map[row["type_code"]]
                 raise NotImplementedError(
                     f"The following Snowflake Data Type is not supported: {error}"
                 )
-            elif row["type_code"] in [1, 2, 3, 4, 6, 7, 8, 10, 11, 13]:
+            elif row["type_code"] in {1, 2, 3, 4, 6, 7, 8, 10, 11, 13}:
```
Comment (sudohainguyen) [inline sdk/python/feast/infra/offline_stores/snowflake_source.py:288]: why change this 🤔

File: sdk/python/feast/infra/offline_stores/snowflake_source.py [hunk 4]
```diff
@@ -279,20 +280,21 @@ def get_table_column_names_and_types(
...
                 row["snowflake_type"] = snowflake_type_code_map[row["type_code"]]
             else:
                 raise NotImplementedError(
                     f"The following Snowflake Column is not supported: {row['column_name']} (type_code: {row['type_code']})"
                 )
 
         return [
-            (column["column_name"], column["snowflake_type"]) for column in metadata
+            (str(column["column_name"]), str(column["snowflake_type"]))
```
Comment (sudohainguyen) [inline sdk/python/feast/infra/offline_stores/snowflake_source.py:296]: do we need to fix one in the PR scope?

File: sdk/python/feast/infra/offline_stores/offline_store.py [hunk 5]
```diff
@@ -146,11 +146,11 @@ def to_arrow(
 
         return pyarrow.Table.from_pandas(features_df)
 
-    def to_sql(self) -> str:
+    def to_sql(self) -> str:  # type: ignore
```
Comment (sudohainguyen) [inline sdk/python/feast/infra/offline_stores/offline_store.py:149]: just wondering is it fine to keep the `raise`?

File: sdk/python/feast/infra/registry/base_registry.py [hunk 6]
```diff
@@ -51,6 +51,7 @@ def apply_entity(self, entity: Entity, project: str, commit: bool = True):
             project: Feast project that this entity belongs to
             commit: Whether the change should be persisted immediately
         """
+        pass
```
Comment (sudohainguyen) [inline sdk/python/feast/infra/registry/base_registry.py:54]: all the `pass` should be changed to
`raise NotImplementedError`

File: sdk/python/feast/repo_config.py [hunk 7]
```diff
@@ -142,27 +137,27 @@ class RepoConfig(FeastBaseModel):
...
-    _online_config: Any = Field(alias="online_store")
+    online_config: Any = Field(None, alias="online_store")
     """ OnlineStoreConfig: Online store configuration (optional depending on provider) """
 
-    _offline_config: Any = Field(alias="offline_store")
+    offline_config: Any = Field(None, alias="offline_store")
     """ OfflineStoreConfig: Offline store configuration (optional depending on provider) """
 
-    _batch_engine_config: Any = Field(alias="batch_engine")
+    batch_engine_config: Any = Field(None, alias="batch_engine")
```
Comment (sudohainguyen) [inline sdk/python/feast/repo_config.py:154]: keep the prefix `_` since it's out of scope

File: sdk/python/feast/repo_config.py [hunk 8]
```diff
@@ -420,15 +409,13 @@ def _validate_feature_server_config(cls, values):
             )
             feature_server_config_class(**values["feature_server"])
         except ValidationError as e:
-            raise ValidationError(
-                [ErrorWrapper(e, loc="feature_server")],
-                model=RepoConfig,
-            )
+            raise e
```
Comment (sudohainguyen) [inline sdk/python/feast/repo_config.py:412]: no need to change this right?

File: sdk/python/tests/integration/feature_repos/universal/data_source_creator.py [hunk 9]
```diff
@@ -42,19 +42,20 @@ def create_data_source(
             A Data source object, pointing to a table or file that is uploaded/persisted for the purpose of the
             test.
         """
-        ...
+        pass
```
Comment (sudohainguyen) [inline sdk/python/tests/integration/feature_repos/universal/data_source_creator.py:45]: `raise NotImplementedError` here as well

File: sdk/python/tests/integration/feature_repos/universal/data_sources/bigquery.py [hunk 10]
```diff
@@ -64,10 +64,10 @@ def create_data_source(
         self,
         df: pd.DataFrame,
         destination_name: str,
-        timestamp_field="ts",
+        event_timestamp_column="ts",
```
Comment (sudohainguyen) [inline sdk/python/tests/integration/feature_repos/universal/data_sources/bigquery.py:67]: `event_timestamp_column` is deprecated isn't it?

File: sdk/python/tests/integration/feature_repos/universal/data_sources/file.py [hunk 11]
```diff
@@ -39,9 +39,10 @@ def create_data_source(
         self,
         df: pd.DataFrame,
         destination_name: str,
-        timestamp_field="ts",
+        event_timestamp_column="ts",
```
Comment (sudohainguyen) [inline sdk/python/tests/integration/feature_repos/universal/data_sources/file.py:42]: `event_timestamp_column` is deprecated isn't it?

File: sdk/python/tests/integration/feature_repos/universal/data_sources/file.py [hunk 12]
```diff
@@ -167,11 +169,11 @@ def _upload_parquet_file(self, df, file_name, minio_endpoint):
     def create_data_source(
         self,
         df: pd.DataFrame,
-        destination_name: Optional[str] = None,
-        suffix: Optional[str] = None,
-        timestamp_field="ts",
+        destination_name: str,
+        event_timestamp_column="ts",
```
Comment (sudohainguyen) [inline sdk/python/tests/integration/feature_repos/universal/data_sources/file.py:173]: here as well

File: sdk/python/tests/unit/cli/test_cli_chdir.py [hunk 13]
```diff
@@ -18,14 +18,16 @@ def test_cli_chdir() -> None:
         repo_path = temp_path / "my_project" / "feature_repo"
         assert result.returncode == 0
 
-        result = runner.run(["--chdir", repo_path, "apply"], cwd=temp_path)
+        result = runner.run(["--chdir", str(repo_path), "apply"], cwd=temp_path)
```
Comment (sudohainguyen) [inline sdk/python/tests/unit/cli/test_cli_chdir.py:21]: should we need casting?

**Review by sudohainguyen — CHANGES_REQUESTED** [review]
comments added

File: sdk/python/feast/infra/passthrough_provider.py (hunk 2 above)
Comment (shuchu) [reply sdk/python/feast/infra/passthrough_provider.py:73]: In Pydantic v2, the variables starting with underscore will be private variables and are not included in the model schema: `Attributes whose name has a leading underscore are not treated as fields by Pydantic, and are not included in the model schema. Instead, these are converted into a "private attribute" which is not validated or even set during calls to __init__, model_validate, etc.`

The original setting (in Pydantic v1), they are supposed to be included in the model schema.

File: sdk/python/feast/infra/offline_stores/snowflake_source.py (hunk 3 above)
Comment (shuchu) [reply sdk/python/feast/infra/offline_stores/snowflake_source.py:288]: just for the running efficiency. :)

File: sdk/python/feast/infra/offline_stores/snowflake_source.py (hunk 4 above)
Comment (shuchu) [reply sdk/python/feast/infra/offline_stores/snowflake_source.py:296]: Yes, it's for Pydantic v2's type check. The return type is "Iterable[Tuple[str, str]]"

File: sdk/python/feast/repo_config.py (hunk 8 above)
Comment (shuchu) [reply sdk/python/feast/repo_config.py:412]: The Pydantic v2 does not have ErrorWrapper() anymore.

File: sdk/python/tests/integration/feature_repos/universal/data_sources/file.py (hunk 11 above)
Comment (shuchu) [reply sdk/python/tests/integration/feature_repos/universal/data_sources/file.py:42]: I added it for passing the tests, although it is not used. Let me see if I can remove it.

File: sdk/python/tests/unit/cli/test_cli_chdir.py (hunk 13 above)
Comment (shuchu) [reply sdk/python/tests/unit/cli/test_cli_chdir.py:21]: Yes, the type of the second argument of runner.run() is "str". The "repo_path" is an object.

**Comment by shuchu** [comment]
> comments added

Thank you and appreciate your time :) let me fix those issues.

File: sdk/python/feast/infra/offline_stores/snowflake_source.py (hunk 3 above)
Comment (shuchu) [reply sdk/python/feast/infra/offline_stores/snowflake_source.py]: changed back.

File: sdk/python/feast/repo_config.py [hunk 14]
```diff
@@ -414,7 +414,6 @@ def _validate_feature_server_config(cls, values: Any) -> Any:
         return values
 
     @field_validator("project")
-    @classmethod
```
Comment (sudohainguyen) [inline sdk/python/feast/repo_config.py]: https://docs.pydantic.dev/2.6/api/functional_validators/#pydantic.functional_validators.model_validator
just checked here, actually we do need `classmethod`

File: sdk/python/feast/repo_config.py [hunk 15]
```diff
@@ -347,14 +342,11 @@ def _validate_online_store_config(cls, values):
...
-            raise ValidationError(
-                [ErrorWrapper(e, loc="online_store")],
-                model=RepoConfig,
-            )
+            raise e
         return values
 
-    @root_validator(pre=True)
-    def _validate_offline_store_config(cls, values):
+    @model_validator(mode="before")
```
Comment (sudohainguyen) [inline sdk/python/feast/repo_config.py:348]: turns out also need to add classmethod for `model_validator` too
https://docs.pydantic.dev/latest/concepts/validators/#model-validators

File: sdk/python/tests/unit/cli/test_cli_chdir.py (hunk 13 above)
Comment (sudohainguyen) [reply sdk/python/tests/unit/cli/test_cli_chdir.py:21]: should we cast once for all from line 18 instead of repeated `str()` calls

**Comment by bushwhackr** [comment]
looks like the test failed due to rate limiting, rerun?

**Pull request merged**